### ISSUE TITLE
Issue using JQuery 3.1.1 and CSP

### ISSUE THREAD

**Issue by victorgaspar** [jquery/jquery#3541]
### Description ###
Hi,

I'm using JQuery version 3.1.1 and I'm trying to implement Content Security Policy directives on my webpage.

I'm getting the following error:
Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self' 'nonce-c20t41c7-73c6-4bf9-fde8-24a7b35t5f71'". Either the 'unsafe-inline' keyword, a hash ('sha256-KAcpKskREkEQf5B3mhDTonpPg34XnzaUC5IoBrOUrwY='), or a nonce ('nonce-...') is required to enable inline execution.

The error is produced on line 82 of the main jquery.js script file. The content of this line is:
doc.head.appendChild( script ).parentNode.removeChild( script );

Basically, it adds an inline script tag to the DOM, and violates the CSP.

I do not want to use 'unsafe-inline'. Is there any other way to circumvent this error? 

As you can see on the CSP violation, I'm using CSP level 2 (nonce). Would it be possible (some how) to inform JQuery to use this nonce when appending the script tag?

Thank you very much.

Kind regards,
V.

**Comment by Brian-Aykut** [comment]
Looks like all attributes get stripped by this call on line 5782 
	`DOMEval( node.textContent.replace( rcleanScript, "" ), doc );`
A new script tag gets created without any attributes and as a result the nonce is no longer present which causes the CSP violation. 

The following can be used to demonstrate/replicate the issue: 

```
<!DOCTYPE html>
<html>
<head>
    <meta http-equiv="Content-Security-Policy" content="default-src http://localhost 'nonce-123456' ; child-src 'none'; object-src 'none'; script-src 'nonce-123456';">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.js" nonce="123456"></script> <!-- HTML nonce works -->
    <script nonce="123456">
        //This works
        console.log('Inline nonce works');

        //This will also work
        var s = document.createElement('script');
        s.setAttribute('nonce', '123456');
        s.innerHTML = 'console.log("Dynamically generated inline tag works")';
        document.head.appendChild(s);

        // This won't work
        var s2 = document.createElement('script');
        s2.setAttribute('nonce', '123456');
        s2.innerHTML = 'console.log("Dynamically generated inline tag appended via jQuery doesn\'t work")';
        $(document.head).append( s2 ); // This will throw a CSP error    
    </script>    
</head>
<body>

</body>
</html>
```

**Comment by timmywil** [comment]
Seems like this would be easy to fix. We can probably add the nonce attribute to our script tag if it exists.

**Comment by Brian-Aykut** [comment]
I've implemented a fix by adding an optional argument to DOMEval which takes an object of attributes and passing the nonce through to it if one is present on a script tag. It currently passes all tests and I've added a relevant test to test that the nonce attribute does remain.

 https://github.com/Brian-Aykut/jquery/tree/3541-csp

Let me know if you'd like me to submit a PR or make any changes :)

**Comment by victorgaspar** [comment]
Hi Brian, your solution works flawlessly! Hopefully it is promoted to main branch soon enough. In the meantime, I'm going to remove JQuery completly and use plain javascript to enable CSP. I'm quite surprise that such a powerful and widely use library as JQuery is not 'CSP friendly'. It took me two days until I finally decided to post the question.

Thanks for the help!

**Comment by koczkatamas** [comment]
Maybe I posting to the wrong issue (fix me pls!), but jQuery fails to handle the most common CSP-related case correctly.

In my case `replaceWith` is called with a HTML including this snippet
```html
<script src="/zerokit/docs/assets/js/page_helper.js"></script>
```

And jQuery takes this script tag, downloads the JS file with Ajax (`evalUrl`) and tries to put the JS code inline into a new `script` tag via which triggers CSP (`unsafe-inline` is disabled in CSP which is basically a must-have configuration).

```javascript
var n = t.createElement("script");
n.text = e,
t.head.appendChild(n).parentNode.removeChild(n)
```

Is there any workaround for this? Should I open a new issue or will this be handled as part of this issue?

**Comment by dmethvin** [comment]
The changes that @Brian-A wrote [here](https://github.com/Brian-A/jquery/commit/a6c07aed9a4e60f653a48b351f59967984ab9562) look good, we just need a PR with that and some unit tests to ensure they work and aren't broken by subsequent changes.

**Comment by Brian-A** [comment]
@dmethvin I've submitted a PR [here](https://github.com/jquery/jquery/pull/3766). 

@koczkatamas if you can include a nonce in your policy & snippet my patch should do the trick (See [above for example](https://github.com/jquery/jquery/issues/3541#issuecomment-282528597)), alternatively if you're able to (which I'm guessing is unlikely or difficult given the details you've provided) the following will work without violating CSP: 

```
var s = $("<script></script>");
$('body').append(s);
s.attr('src', 'test.js');
```

Hopefully this is helpful :)

**Comment by subversivo58** [comment]
Accompanying the **[PR](https://github.com/jquery/jquery/pull/3766)...** was there any definition about the change request?

The good application of CSP is important for the application, is there any forecast of when this can be added to the core?

**Comment by Brian-A** [comment]
I updated the PR with the requested changes although I see now it may have gotten missed/lost due to the thread being marked as outdated? https://github.com/jquery/jquery/pull/3766#discussion_r138805854

**Comment by cnizzardini** [comment]
Has this issue been resolved in any versions of jQuery?

**Comment by timmywil** [comment]
Not yet, but this should make it into the next release.

**Comment by mgol** [comment]
@timmywil If that's the case, the Milestone should get updated. Now it just points to `Future`.

**Comment by mgol** [comment]
I've been wondering - the error is caused by us manually evaling the downloaded script. Now, when cross domain we already do it differently, via a script tag with the src attribute set, see:
https://github.com/jquery/jquery/blob/3.3.1/src/ajax/script.js#L47-L74

Is there any reason why this has to be restricted to cross-domain?

**Comment by dmethvin** [comment]
It's been a LOOOONG time but I think this was because the onerror event for scripts is relatively recent--is it now supported everywhere? To avoid silent failures for 404s and script errors we grabbed the source and evaled it.

**Comment by mgol** [comment]
@dmethvin I've just tested (test case: http://output.jsbin.com/kosexoy) and it doesn't work in IE 8 but works in IE 9+; it also works in iOS 7+ which is what we support.

Unfortunately, it also doesn't work in Android Browser (all 4.0-4.3). Fortunately, Android Browser doesn't seem to support CSP so perhaps it'd work to extend the cross-domain way of loading scripts to modern browsers? Not sure how to check it, though. Android Browser satisfies `document.createElement('script').onerror === null` even though unknown attributes should have a value of `undefined`.

**Comment by aimee-katherine** [comment]
Hi, I'm using JQuery version 3.3.1 and I'm trying to implement Content Security Policy on my webpage.

I'm getting the following error:

Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self' 'nonce-8WJnAqGXkJewcnh5byFtLu1IRq9Zu0Adv31B0k18DLaNBD1xHb19y8ppRMlY'
The error is produced on line 2 of jquery.min.js script file. The content of this line is:

function m(e,t,n){var i,o=(t=t||r).createElement("script");
Is there a way around this problem without having to use "unsafe-inline"?  I've been looking all over the internet, desperate for a solution.  =(  I tried what Brian suggested to no avail...

**Comment by Brian-A** [comment]
Hi @aimee-katherine you could try adding a hash ( sha256-yourhash; )of the script to your CSP although this isn't something I've tested with dynamically generated script tags I only tested if the attributes on the tag were stripped. Alternatively I've patched 3.1.1 to work with some dynamically generated tags and should work fine if the tag is generated using jQuery and a nonce (or possibly hash) is applied https://github.com/Brian-A/jquery/tree/3541-csp 

If my comments don't help an example of how you're creating the script tags or a test case would help immensely :)

**Comment by Pomax** [comment]
@mgol on a technical note, IE8 and IE9 were finally fully discontinued by Microsoft in mid-January of 2016, so it might not make sense to still aim for compat in these browsers anymore.

**Comment by mgol** [comment]
@Pomax The same applies to IE 10. That said, jQuery wouldn't gain a lot in size by cutting just IE 9 & 10 support; even cutting out IE 6-8 support only brought a ~12% size decrease and it eliminated a whole alternative event model & a lot of quirks. Apart from size a good motivation to dropping support would be an issue that is hard to fix while maintaining IE <11 support.

We might drop IE <11 one day but since most JS frameworks & (non-UI) libraries still support IE 9} I don't think we should be the ones to cut support first.

UI libraries like jQuery UI are dropping IE <11 a lot since they want to rely on flexbox; this is not a concern for jQuery core.

**Comment by graingert** [comment]
And IE6 is still supported on ATMs

**Comment by mgol** [comment]
@graingert Is it? What's the source of this information? The official Microsoft IE support policy document claims IE<8 isn't supported anywhere anymore: https://support.microsoft.com/en-us/help/17454/lifecycle-faq-internet-explorer

**Comment by graingert** [comment]
woops, I was thinking of IE8

**Comment by jfoclpf** [comment]
Hi, same issue here with **jquery 3.3.1**

I load the jquery as documentation suggests

```
<script
  src="https://code.jquery.com/jquery-3.3.1.min.js"
  integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8="
  crossorigin="anonymous"></script>
```

And in CSP policy I obviously add

`script-src 'self' *.jquery.com`

Nonetheless:

Either the 'unsafe-inline' keyword, a hash ('sha256-hIfJ4PBJkqOTEex5dr9/IAbF6BUlKRcQCnan+wnJkzI='), or a nonce ('nonce-...') is required to enable inline execution.

**Comment by Pomax** [comment]
Is that from "somewhere in your code" (e.g. you're calling loadScript) or is that just from trying to load the base .js file?

**Comment by jfoclpf** [comment]
Hi, thanks for replying. It is really the base file which I load at the end
of html. It seems jQuery has some code built-in that violates such CSP rule.

**Comment by dmethvin** [comment]
@jfoclpf This page works correctly for me. If you have a large file it probably has some inline script hiding in it. The error message you get will provide the error line, you can set a breakpoint on it to see where the stack leads. Please don't continue the debugging session in this ticket though, it's not related except that both involve a CSP message.
```html
<!DOCTYPE html>
<html><head>
    <meta charset="utf-8">
    <meta http-equiv="Content-Security-Policy" content="script-src 'self' code.jquery.com">
    <script src="http://code.jquery.com/jquery-3.3.1.js"></script>
    <title>CSP</title>
</head><body></body></html>
```

**Comment by Pomax** [comment]
As long as your CSP for `script-src` contains `http://code.jquery.com/jquery-3.3.1.js`, you should be able to load it with a plain `<script src="http://code.jquery.com/jquery-3.3.1.js"></script>` tag, without the integrity and crossorigin attributes.

**Comment by jfoclpf** [comment]
Thank you @dmethvin and @Pomax 
Just one more question. I use a lot the jquery function [`$.getScript`](https://api.jquery.com/jquery.getscript/) to load external scripts which are hosted on the _same_ domain. Does this jquery function trigger perhaps this CSP `'unsafe-inline'` violation?

**Comment by dmethvin** [comment]
@jfoclpf yes that would cause it. See gh-3969.

**Comment by cnsgithub** [comment]
Any progress on this? It stops me from integrating Content-Security-Policy in the PrimeFaces JSF library, see https://github.com/primefaces/primefaces/pull/3884.

**Comment by Pomax** [comment]
As long as your CSP headers include http://code.jquery.com/jquery-3.3.1.js in the script-src, you should be able to load it with a plain <script src="http://code.jquery.com/jquery-3.3.1.js"></script> tag, without any need for integrity and crossorigin attributes.

**Comment by graingert** [comment]
@Pomax please read the rest of the issue discussion

**Comment by fr4x1nu5** [comment]
@Pomax That's not the problem at all. The problem is that jquery's `replaceWith` function adds a script to the DOM without taking previously existing `nonce` attributes into account... Btw, as a workaround one might use `document.getElementById(<id>).outerHTML = newHTML` where `newHTML="<script nonce='123' />`.

**Comment by fr4x1nu5** [comment]
Losing the nonce attribute is probably not the only problem associated with it. There might be usecases where non-javascript (the default if `type `attribute is omitted) `script` elements should be replaced, e.g. `type="x-shader/x-vertex"` for WebGL.

**Comment by timmywil** [comment]
We landed another PR that will make solving this pretty much a one-liner. The PR that originally addressed this needs updating. See https://github.com/jquery/jquery/pull/3766.

**Comment by cnsgithub** [comment]
@timmywil Nice, but if I understand it correctly, PR #3766 will only handle the nonce attribute. But what about other attributes as mentioned by @fr4x1nu5?

**Comment by timmywil** [comment]
@cnsgithub Sorry, can you be more specific about which attributes you mean? I must be glazing over som comment from @fr4x1nu5.

**Comment by fr4x1nu5** [comment]
@timmywil As I said, other attributes like `type`, e.g. `type="x-vertex/x-shader"` (for WebGL) or `async` or `defer`...

**Comment by timmywil** [comment]
Okay, I saw your mention of `type`, but I don't see mentions of `async`/`defer`. While these attributes are out of scope for this issue, which is just about CSP and addressed by `nonce`, note that `type` is already covered in `DOMEval.js`. As for the other 2, I don't they're necessary since if you're using the script transport, you already have control over that behavior.

**Comment by cnsgithub** [comment]
@fr4x1nu5 

> Btw, as a workaround one might use document.getElementById(<id>).outerHTML = newHTML where newHTML="<script nonce='123' />.

This will not work as expected since the script will not be executed. I think the best solution is adding `nonce `to `preservedScriptAttributes `as suggested by @timmywil.

**Comment by jlherren** [comment]
I implemented @timmywil's solution by adding `'nonce'` to `preservedScriptAttributes` which works great in Chrome, but still failed on Edge/Firefox. The problem is these browsers seem to be picky about how you read/write the nonce attribute of the script element. So in addition to that I had to modify the code as follows:

```
for ( i in preservedScriptAttributes ) {
	// "node.getAttribute(i)" works in Edge/FF but not in chrome and "node[i]" works in chrome but not in Edge/FF
	// so just try both
	var val = node[i] || node.getAttribute(i);
	if ( val ) {
		// "script[i] = val" doesn't work in Edge/FF, setAttribute() works everywhere
		script.setAttribute(i, val);
	}
}
```

**Comment by boompig** [comment]
What is the current workaround while we wait for the PR?

The issue has been open over 1.5 years so I feel we might be waiting a while longer for the fix to land...

**Comment by timmywil** [comment]
@boompig The issue has a PR that should make it into the next release, which we're trying to close out in the coming weeks. Issues can be open longer than 1.5 years and we purposefully avoid frequent releases. If you need to add a script tag with a noonce attribute, this is pretty easy to do with vanilla JS.

**Comment by mgol** [comment]
A new PR: #4269.

**Comment by tandraschko** [comment]
@mgol do you have a compiled jquery.js of your branch? So i could just test it.

**Comment by mgol** [comment]
@tandraschko The PR has landed so you can try https://code.jquery.com/jquery-git.js. Just please don't load this file in production. :)

**Comment by tandraschko** [comment]
@mgol all right, thanks :) 
What about $.globalEval? Should we provide a way to pass the nonce?
How can i use DOMEval from my scripts until we we introduce something for $.globalEval?
XXX(?).DOMEval(myScript, document, { nonce: 'mynonce') }?

**Comment by mgol** [comment]
@tandraschko It's not possible to pass `nonce` through `$.globalEval` at the moment. The merged PR just makes `nonce` preserved during DOM manipulation. If you'd like to see support for `nonce` in `$.globalEval`, please submit a separate issue. Thanks!

**Comment by tandraschko** [comment]
@mgol Done: https://github.com/jquery/jquery/issues/4278 - Thanks :)

**Comment by jfoclpf** [comment]
Good day
Considering that the error still persists, why the [milestone 18 for v.3.4.0](https://github.com/jquery/jquery/milestone/18) and attached to this issue, doesn't mention this issue?
Thanks in advance
